internal/engine/upper_test.go

```diff
@@ -272,16 +283,41 @@ func (b *fakeBuildAndDeployer) BuildAndDeploy(ctx context.Context, st store.RSto
 		b.resultsByID[key] = val
 	}
 
+	// block until we know we're supposed to resolve this build
+	for {
+		if _, ok := b.buildsAwaitingResolution[count]; ok {
```

Comment:
did you intend to also lock the mutex here? otherwise you'll have race conditions

i might also break this into a waitUntilBuildResolves(index) helper, just to cut down on the size of this method

Reply:
good catch! and that makes sense, will do